File: 19-10-01-elevator/java/src/main/java/elevator/Elevator.java
package elevator;

import java.util.HashSet;
import java.util.Set;

public class Elevator {

    private int floor;
    private int totalTime;

    private Building building;
    private Set<Passenger> passengers;

    public Elevator(Building building) {
        if (building == null) {
            throw new IllegalArgumentException("Building must not be null");
        }

        floor = 0;
        this.building = building;
        passengers = new HashSet<>();
    }

    public void up() {
        if (floor == Integer.MAX_VALUE) {
            return;
        }
        floor++;
        totalTime++;

        changePassengerFloors();
    }

    public void down() {
        if (floor == Integer.MIN_VALUE) {
            return;
        }
        floor--;
        totalTime++;

        changePassengerFloors();
    }

    public void openDoors() {
        totalTime++;

        addPassengers();
        removePassengers();
    }

    private void addPassengers() {
        for (Passenger p : building.getPassengersOnFloor(floor)) {
            if (p.getCurrentFloor() != p.getDestinationFloor()) {
                passengers.add(p);
            }
        }
    }

    private void removePassengers() {
        passengers.removeIf(Passenger::arrived);
    }

    private void changePassengerFloors() {
        for (Passenger p : passengers) {
            p.setCurrentFloor(floor);
        }
    }

    public int getFloor() {
        return floor;
    }

    public int getTotalTime() {
        return totalTime;
    }

    public Set<Passenger> getPassengers() {
        return passengers;
    }

}

File: 19-10-01-elevator/python/Elevator.py
class Elevator:

    floor = 0
    totalTime = 0
    building = None
    passengers = None

    MAX_SIZE = (2**31) - 1
    MIN_SIZE = -(2**31)

    def __init__(self, building):
        if building is None:
            raise Exception("Building cannot be None")

        self.floor = 0
        self.building = building
        self.passengers = []

    def up(self):
        if self.floor == self.MAX_SIZE:
            return
        self.floor += 1
        self.totalTime += 1
        self.changePassengerFloors()

    def down(self):
        if self.floor == self.MIN_SIZE:
            return
        self.floor += 1
        self.totalTime += 1
        self.changePassengerFloors()

    def openDoors(self):
        self.totalTime += 1
        self.addPassengers()
        self.removePassengers()

    def addPassengers(self):
        for p in self.building.getPassengersOnFloor(self.floor):
            if p.getCurrentFloor() != p.destinationFloor():
                self.passengers.append(p)

    def removePassengers(self):
        for p in self.passengers:
            if p.arrived:
                self.passengers.remove(p)

    def changePassengerFloors(self):
        for p in self.passengers:
            p.setCurrentFloor(self.floor)

    def getFloor(self):
        return self.floor

    def getTotalTime(self):
        return self.totalTime

    def getPassengers(self):
        return self.passengers